docs/configuration.md
# Configuration

System-wide configuration parameters can be set by creating a ConfigMap with name `model-serving-config`. It should contain a single key named `config.yaml`, whose value is a yaml doc containing the configuration. All parameters have defaults and are optional. If the ConfigMap does not exist, all parameters will take their defaults.

The configuration can be updated at runtime and will take effect immediately. Be aware however that certain changes could cause temporary disruption to the service - in particular changing the service name, port, TLS configuration and/or headlessness.

Example:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: model-serving-config
data:
  config.yaml: |
    #Sample config overrides
    inferenceServiceName: "modelmesh-serving"
    inferenceServicePort: 8033
    podsPerRuntime: 2
    metrics:
      enabled: true
```

The following parameters are currently supported. _Note_ the keys are expressed here in camel case but are in fact case-insensitive.

| Variable                         | Description                                                                                           | Default             |
| -------------------------------- | ----------------------------------------------------------------------------------------------------- | ------------------- |
| `inferenceServiceName`           | The service name which is used for communication with the serving server                              | `modelmesh-serving` |
| `inferenceServicePort`           | The port number for communication with the inferencing service                                        | `8033`              |
| `storageSecretName`              | The secret containing entries for each storage backend from which models can be loaded (\* see below) | `storage-config`    |
| `podsPerRuntime`                 | Number of server Pods to run per enabled Serving Runtime (\*\* see below)                             | `2`                 |
| `tls.secretName`                 | Kubernetes TLS type secret to use for securing the Service; no TLS if empty (\*\*\* see below)        |                     |
| `tls.clientAuth`                 | Enables mutual TLS authentication. Supported values are `required` and `optional`, disabled if empty  |                     |
| `headlessService`                | Whether the Service should be headless (recommended)                                                  | `true`              |
| `enableAccessLogging`            | Enables logging of each request to the model server                                                   | `false`             |
| `serviceAccountName`             | The service account to use for runtime Pods                                                           | `modelmesh`         |
| `metrics.enabled`                | Enables serving of Prometheus metrics                                                                 | `true`              |
| `metrics.port`                   | Port on which to serve metrics via the `/metrics` endpoint                                            | `2112`              |
| `metrics.scheme`                 | Scheme to use for the `/metrics` endpoint (`http` or `https`)                                         | `https`             |
| `scaleToZero.enabled`            | Whether to scale down Serving Runtimes that have no Predictors                                        | `true`              |
| `scaleToZero.gracePeriodSeconds` | The number of seconds to wait after Predictors are deleted before scaling to zero                     | `60`                |

(\*) Currently requires a controller restart to take effect

(\*\*) This parameter will likely be removed in a future release; the Pod replica counts will become more dynamic.

(\*\*\*) The TLS configuration secret allows for keys:

- `tls.crt` - path to TLS secret certificate
- `tls.key` - path to TLS secret key
- `ca.crt` (optional) - single path or comma-separated list of paths to trusted certificates

## Generating TLS Certificates for Dev/Test

TLS is enabled through adding a value for `tls.secretName` in the user's ConfigMap that points to an existing kube secret with TLS key/cert details.

To create a SAN key/cert for TLS, use command:

```sh
$ openssl req -x509 -newkey rsa:4096 -sha256 -days 3560 -nodes -keyout example.key -out example.crt -subj '/CN=modelmesh-serving' -extensions san -config openssl-san.config
```

Where the contents of `openssl-san.config` look like:

```
[ req ]
distinguished_name = req
[ san ]
subjectAltName = DNS:modelmesh-serving,DNS:localhost,IP:0.0.0.0
```

With the generated key/cert, create a kube secret with contents like:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: <NAME_OF_SECRET>
type: kubernetes.io/tls
stringData:
  tls.crt: <contents-of-example.crt>
  tls.key: <contents-of-example.key>
  ca.crt: <contents-of-example.crt>
```

For basic TLS, only the fields `tls.crt` and `tls.key` are needed in the kube secret. For mutual TLS, add `ca.crt` in the kube secret and set the configuration `tls.clientAuth` to `require` in the ConfigMap `model-serving-config`.


docs/install/environment.md
## Prerequisites

- **Kubernetes cluster** - A Kubernetes cluster is required. You will need `cluster-admin` authority in order to complete all of the prescribed steps.

- **Kubectl and Kustomize** - The installation will occur via the terminal using [kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl) and [kustomize](https://kubectl.docs.kubernetes.io/installation/kustomize/).

- **etcd** - ModelMesh Serving requires an [etcd](https://etcd.io/) server in order to coordinate internal state which can be either dedicated or shared. More on this later.

We provide an install script to quickly run ModelMesh Serving with a provisioned etcd server. This may be useful for experimentation or development but should not be used in production.

## Namespace Scope

ModelMesh Serving is namespace scoped, meaning all of its components must exist within a single namespace and only one instance of ModelMesh Serving can be installed per namespace.Multiple ModelMesh Serving instances can be installed in separate namespaces within the cluster.

## Deployed Components

|            | Type             | Pod                        | Count | Default CPU request/limit per-pod | Default mem request/limit per-pod |
| ---------- | ---------------- | -------------------------- | ----- | --------------------------------- | --------------------------------- |
| 1          | Controller       | modelmesh controller pod   | 1     | 50m / 1                           | 96Mi / 512Mi                      |
| 2          | Object Storage   | minio pod                  | 1     | 200m / 200m                       | 256Mi / 256Mi                     |
| 3          | Metastore        | ETCD pods                  | 1     | 200m / 200m                       | 512Mi / 512Mi                     |
| 4          | Built-in Runtime | Nvidia Triton runtime Pods | 2     | 850m / 10                         | 1568Mi / 1984Mi                   |
| 5          | Built-in Runtime | The MLServer runtime Pods  | 2     | 850m / 10                         | 1568Mi / 1984Mi                   |
| **totals** |                  |                            | 7     | 3850m / 41.4                      | 6.96Gi / 8.59Gi                   |

When ModelMesh serving instance is installed with the `--quickstart` option, pods shown in 1 - 5 are created with a total CPU(request/limit) of 3850m / 41.4 and total memory(request/limit) of 6.96Gi / 8.59Gi.

By default, runtime Pods will only be started when there's at least one Predictor created that uses a corresponding model type.

The deployed footprint can be significantly reduced in the following ways:

- Individual built-in runtimes can be disabled by setting `disabled: true` in their corresponding `ServingRuntime` resource - if the corresponding model types aren't used.

- The number of Pods per runtime can be changed from the default of 2 (e.g. down to 1), via the `podsPerRuntime` global configuration parameter (see [configuration](../configuration)). It is recommended for this value to be a minimum of 2 for production deployments.

- Memory and/or CPU resource allocations can be reduced (or increased) on the primary model server container in either of the built-in `ServingRuntime` resources (container name `triton` or `mlserver`). This has the effect of adjusting the total capacity available for holding served models in memory.

```shell
> kubectl edit servingruntime triton-2.x
> kubectl edit servingruntime mlserver-0.x
```

Please be aware that:

1. Changes made to the _built-in_ runtime resources will likely be reverted when upgrading/re-installing
2. Most of this resource allocation behaviour/config will change in future versions to become more dynamic - both the number of pods deployed and the system resources allocated to them

The following resources will also be created in the same namespace:

- `model-serving-defaults` - ConfigMap holding default values tied to a release, should not be modified. Configuration can be overriden by creating a user ConfigMap, see [configuration](../configuration)
- `tc-config` - ConfigMap used for some internal coordination
- `storage-config` - Secret holding config for each of the storage backends from which models can be loaded - see [storage config[(../predictors/storage.md) and [the example](../predictors/README.md)


controllers/config.go
// Copyright 2021 IBM Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package controllers

import (
	"context"
	"fmt"
	"os"
	"path"
	"runtime"
	"strings"
	"sync"
	"sync/atomic"
	"unsafe"

	"sigs.k8s.io/controller-runtime/pkg/handler"
	"sigs.k8s.io/controller-runtime/pkg/reconcile"

	"github.com/spf13/viper"
	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/errors"
	"k8s.io/apimachinery/pkg/api/resource"
	"k8s.io/apimachinery/pkg/types"
	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/client"
)

const (
	EnvEtcdSecretName     = "ETCD_SECRET_NAME"
	DefaultEtcdSecretName = "model-serving-etcd"
)

var (
	defaultConfig *viper.Viper
	configLog     = ctrl.Log.WithName("config")
)

// Type Config holds process global configuration information
type Config struct {
	// System config
	EtcdSecretName    string // DEPRECATED - should be removed in the future
	ModelMeshEndpoint string // For dev use only

	// Service config
	InferenceServiceName    string
	InferenceServicePort    uint16
	TLS                     TLSConfig
	HeadlessService         bool
	GrpcMaxMessageSizeBytes int

	// Runtimes config
	ModelMeshImage         ImageConfig
	ModelMeshResources     ResourceRequirements
	StorageHelperImage     ImageConfig
	StorageHelperResources ResourceRequirements
	PodsPerRuntime         uint16
	StorageSecretName      string
	EnableAccessLogging    bool

	ServiceAccountName string

	Metrics     PrometheusConfig
	ScaleToZero ScaleToZeroConfig
}

type PrometheusConfig struct {
	Enabled                          bool
	Port                             uint16
	Scheme                           string
	DisablePrometheusOperatorSupport bool
}

type ScaleToZeroConfig struct {
	Enabled bool
	// how long to wait after the last predictor assigned to a runtime is deleted
	// before scaling to zero
	GracePeriodSeconds uint16
}

type TLSConfig struct {
	// TLS disabled if omitted
	SecretName string
	// Mutual TLS disabled if omitted
	ClientAuth string
}

func (c *Config) GetEtcdSecretName() string {
	secretName, found := os.LookupEnv(EnvEtcdSecretName)
	if !found {
		secretName = DefaultEtcdSecretName
	}

	// for backward compatability with the old configmap - should be removed in the future
	if secretName == DefaultEtcdSecretName && c.EtcdSecretName != "" {
		secretName = c.EtcdSecretName
	}

	return secretName
}

// Type Config holds process global configuration information
type ConfigProvider struct {
	config                unsafe.Pointer
	c                     sync.Cond
	isReloading           bool
	loadedResourceVersion string
}

func NewConfigProvider(ctx context.Context, cl client.Client, name types.NamespacedName) (*ConfigProvider, error) {
	// Perform initial load of the default configuration
	config, err := NewMergedConfigFromString("")
	if err != nil {
		return nil, err
	}

	cp := &ConfigProvider{c: sync.Cond{L: &sync.Mutex{}}, config: (unsafe.Pointer)(config)}
	return cp, cp.ReloadConfigMap(ctx, cl, name)
}

func (cp *ConfigProvider) GetConfig() *Config {
	return (*Config)(atomic.LoadPointer(&cp.config))
}

func (cp *ConfigProvider) ReloadConfigMap(ctx context.Context, c client.Client, name types.NamespacedName) error {
	cp.c.L.Lock()
	cp.isReloading = true
	defer func() {
		cp.c.L.Unlock()
		cp.isReloading = false
	}()

	// get the user ConfigMap
	configmap := corev1.ConfigMap{}
	err := c.Get(ctx, name, &configmap)
	// any error other than NotFound is unexpected
	if err != nil && !errors.IsNotFound(err) {
		return err
	}

	// only reload if the resource version changed from what we last loaded successfully
	//  if the user config does not exist, configmap.ResourceVersion is the empty string
	if configmap.ResourceVersion != cp.loadedResourceVersion {
		var err2 error
		var newConfig *Config
		//  if the configmap was deleted but had been previously loaded
		if configmap.ResourceVersion == "" {
			configLog.Info("User configmap deleted, reverting to defaults", "ConfigMap", name)
			// load from empty string to by-pass validation of the configmap
			newConfig, err2 = NewMergedConfigFromString("")
		} else {
			configLog.Info("Reloading user config", "ConfigMap", name)
			newConfig, err2 = NewMergedConfigFromConfigMap(configmap)
		}
		if err2 != nil {
			return err2
		}
		// update the stored resource version to track changes
		atomic.StorePointer(&cp.config, (unsafe.Pointer)(newConfig))
		cp.loadedResourceVersion = configmap.ResourceVersion
	}

	cp.c.Broadcast()

	return nil
}

// Handler used by controllers which depend on the user configuration
func ConfigWatchHandler(configMapName types.NamespacedName, f func() []reconcile.Request,
	cp *ConfigProvider, kclient *client.Client) handler.EventHandler {
	return handler.EnqueueRequestsFromMapFunc(func(o client.Object) []reconcile.Request {
		// Ignore ConfigMaps we don't care about
		if o.GetName() == configMapName.Name && o.GetNamespace() == configMapName.Namespace {
			err := cp.ReloadConfigMap(context.TODO(), *kclient, configMapName)
			if err != nil {
				configLog.Error(err, "Unable to reload user configuration")
			}
			return f()
		}
		return []reconcile.Request{}
	})
}

func (cp *ConfigProvider) IsReloading() bool {
	return cp.isReloading
}

func (cp *ConfigProvider) AwaitReload() {
	cp.c.L.Lock()
	defer cp.c.L.Unlock()

	if cp.isReloading {
		cp.c.Wait()
	}
}

type ImageConfig struct {
	Name    string
	Tag     string
	Digest  string
	Command []string
}

func (m ImageConfig) TaggedImage() string {
	// Use only the digest if it is provided and ignore the tag.
	// It is possible to reference an image by both tag and digest, but this is considered ambiguous and the digest would be used to pull the image anyway.
	// See also https://github.com/cri-o/cri-o/pull/3060
	if m.Digest != "" {
		return fmt.Sprintf("%s@%s", m.Name, m.Digest)
	} else if m.Tag != "" {
		return fmt.Sprintf("%s:%s", m.Name, m.Tag)
	}
	return m.Name
}

type ResourceRequirements struct {
	Requests ResourceQuantities
	Limits   ResourceQuantities
	// used to cache the parsed resources from parseAndValidate()
	parsedKubeResourceRequirements *corev1.ResourceRequirements
}
type ResourceQuantities struct {
	CPU    string
	Memory string
}

func (rr ResourceRequirements) ToKubernetesType() *corev1.ResourceRequirements {
	// assert that parseAndValidate has already been called
	if rr.parsedKubeResourceRequirements == nil {
		panic("ResourceRequirements: Must call parseAndValidate() before ToKubernetesType()")
	}

	return rr.parsedKubeResourceRequirements
}

func (rr *ResourceRequirements) parseAndValidate() error {
	var err error
	var limitsCPU, requestsCPU, limitsMemory, requestsMemory resource.Quantity

	if limitsCPU, err = resource.ParseQuantity(rr.Limits.CPU); err != nil {
		return fmt.Errorf("Cannot parse 'Limits.CPU' as a quantity: %w", err)
	}
	if requestsCPU, err = resource.ParseQuantity(rr.Requests.CPU); err != nil {
		return fmt.Errorf("Cannot parse 'Requests.CPU' a quantity: %w", err)
	}

	if limitsMemory, err = resource.ParseQuantity(rr.Limits.Memory); err != nil {
		return fmt.Errorf("Cannot parse 'Limits.Memory' as a quantity: %w", err)
	}
	if requestsMemory, err = resource.ParseQuantity(rr.Requests.Memory); err != nil {
		return fmt.Errorf("Cannot parse 'Requests.Memory' a quantity: %w", err)
	}

	rr.parsedKubeResourceRequirements = &corev1.ResourceRequirements{
		Limits: corev1.ResourceList{
			corev1.ResourceCPU:    limitsCPU,
			corev1.ResourceMemory: limitsMemory,
		},
		Requests: corev1.ResourceList{
			corev1.ResourceCPU:    requestsCPU,
			corev1.ResourceMemory: requestsMemory,
		},
	}

	return nil
}

// Sets the defaults prior to config file parsing
// Note: see also config/default/defaults.yaml for the shipped default config
func defaults(v *viper.Viper) {
	v.SetDefault("InferenceServiceName", "modelmesh-serving")
	v.SetDefault("InferenceServicePort", 8033)
	v.SetDefault("PodsPerRuntime", 2)
	v.SetDefault("StorageSecretName", "storage-config")
	v.SetDefault("ServiceAccountName", "modelmesh")
	v.SetDefault("Metrics.Port", 2112)
	v.SetDefault("Metrics.Scheme", "https")
	v.SetDefault("ScaleToZero.Enabled", true)
	v.SetDefault("ScaleToZero.GracePeriodSeconds", 60)
	// default size 16MiB in bytes
	v.SetDefault("GrpcMaxMessageSizeBytes", 16777216)
}

func init() {
	defaultConfig = viper.New()

	defaults(defaultConfig)

	defaultConfig.SetConfigName("config-defaults")
	defaultConfig.SetConfigType(ConfigType)
	defaultConfig.AddConfigPath(MountLocation)

	//For dev env and tests, must get the config filename using cwd relative path
	if _, filename, _, ok := runtime.Caller(0); !ok {
		panic("Unable to get the caller")
	} else {
		filepath := path.Join(path.Dir(filename), "../config/default")
		defaultConfig.AddConfigPath(filepath)
	}

	if err := defaultConfig.ReadInConfig(); err != nil {
		configLog.Error(err, "Unable to read the default configuration", "path", MountLocation)
	}
}

func NewMergedConfigFromConfigMap(m corev1.ConfigMap) (*Config, error) {
	configYaml, ok := m.Data["config.yaml"]
	if !ok {
		err := fmt.Errorf("User ConfigMap must contain a key named config.yaml")
		return nil, err
	}

	return NewMergedConfigFromString(configYaml)
}

func NewMergedConfigFromString(configYaml string) (*Config, error) {
	var err error

	v := viper.New()
	v.SetConfigType(ConfigType)
	for _, key := range defaultConfig.AllKeys() {
		v.SetDefault(key, defaultConfig.Get(key))
	}

	configYamlReader := strings.NewReader(configYaml)
	if err = v.ReadConfig(configYamlReader); err != nil {
		return nil, err
	}

	// Even if the default config has an image digest, a user should be able to
	// override it with a tag (ignoring the default digest)
	// HACK: There should be a better way to do this...
	if v.GetString("storageHelperImage.tag") != defaultConfig.GetString("storageHelperImage.tag") &&
		v.GetString("storageHelperImage.digest") == defaultConfig.GetString("storageHelperImage.digest") {
		v.Set("storageHelperImage.digest", "")
	}
	if v.GetString("modelMeshImage.tag") != defaultConfig.GetString("modelMeshImage.tag") &&
		v.GetString("modelMeshImage.digest") == defaultConfig.GetString("modelMeshImage.digest") {
		v.Set("modelMeshImage.digest", "")
	}

	// unmarshal the config into a Config struct
	var config Config
	if err = v.Unmarshal(&config); err != nil {
		return nil, err
	}

	configLog.Info("Updated model serving config", "mergedConfig", config)

	// extra validations on parsed config
	if err = config.ModelMeshResources.parseAndValidate(); err != nil {
		return nil, fmt.Errorf("Invalid config for 'ModelMeshResources': %s", err)
	}
	if err = config.StorageHelperResources.parseAndValidate(); err != nil {
		return nil, fmt.Errorf("Invalid config for 'StorageHelperResources': %s", err)
	}

	return &config, nil
}
